File: frontend/src/lib/calculations.js
/**
 * FINWISE Financial Calculations
 *
 * This module mirrors financial/calculations.py and financial/what_if.py
 * exactly. All arithmetic here must stay in sync with the Python engine.
 * When the FastAPI backend is connected, this is the fallback/offline layer.
 */

/**
 * Calculate core monthly financial metrics.
 * Mirrors: calculate_financial_snapshot()
 */
export function calculateFinancialSnapshot({
  income,
  housing,
  food,
  transportation,
  education,
  healthcare,
  entertainment,
  subscriptions,
  other,
  debtPayment,
  savingsContribution,
}) {
  const totalExpenses =
    housing +
    food +
    transportation +
    education +
    healthcare +
    entertainment +
    subscriptions +
    other +
    debtPayment;

  const monthlySurplus = income - totalExpenses;

  const savingsRate =
    income > 0 ? (savingsContribution / income) * 100 : null;

  const debtToIncome =
    income > 0 ? (debtPayment / income) * 100 : null;

  const surplusRate =
    income > 0 ? (monthlySurplus / income) * 100 : null;

  return {
    totalMonthlyExpenses: totalExpenses,
    monthlySurplus,
    cashFlowStatus: monthlySurplus >= 0 ? 'surplus' : 'deficit',
    savingsRate,
    surplusRate,
    debtToIncomeRatio: debtToIncome,
  };
}

/**
 * Calculate progress toward a financial goal.
 * Mirrors: calculate_goal_progress()
 */
export function calculateGoalProgress({ targetAmount, amountSaved }) {
  if (targetAmount <= 0) {
    return { remainingAmount: 0, progressPercentage: 100 };
  }

  const remainingAmount = Math.max(targetAmount - amountSaved, 0);
  const progressPercentage = Math.min((amountSaved / targetAmount) * 100, 100);

  return { remainingAmount, progressPercentage };
}

/**
 * Calculate how much needs to be saved monthly to hit a goal by a date.
 * Mirrors: calculate_goal_plan()
 */
export function calculateGoalPlan({
  targetAmount,
  amountSaved,
  monthlySavings,
  targetDate,
}) {
  const remainingAmount = Math.max(targetAmount - amountSaved, 0);

  if (remainingAmount === 0) {
    return {
      remainingAmount: 0,
      requiredMonthlySaving: 0,
      currentSavingSufficient: true,
      monthsRemaining: 0,
    };
  }

  if (monthlySavings <= 0) {
    return {
      remainingAmount,
      requiredMonthlySaving: null,
      currentSavingSufficient: false,
      monthsRemaining: null,
    };
  }

  const today = new Date();
  const target = new Date(targetDate);

  let months =
    (target.getFullYear() - today.getFullYear()) * 12 +
    (target.getMonth() - today.getMonth());

  if (target.getDate() >= today.getDate()) {
    months += 1;
  }

  if (months <= 0) {
    return {
      remainingAmount,
      requiredMonthlySaving: null,
      currentSavingSufficient: false,
      monthsRemaining: 0,
    };
  }

  const requiredMonthlySaving = remainingAmount / months;

  return {
    remainingAmount,
    requiredMonthlySaving,
    currentSavingSufficient: monthlySavings >= requiredMonthlySaving,
    monthsRemaining: months,
  };
}

/**
 * Calculate the impact of a what-if scenario.
 * Mirrors: calculate_what_if()
 */
export function calculateWhatIf({
  income, housing, food, transportation, education,
  healthcare, entertainment, subscriptions, other,
  debtPayment, savingsContribution,
  // deltas
  incomeDelta = 0,
  housingDelta = 0,
  foodDelta = 0,
  transportationDelta = 0,
  educationDelta = 0,
  healthcareDelta = 0,
  entertainmentDelta = 0,
  subscriptionsDelta = 0,
  otherDelta = 0,
  debtPaymentDelta = 0,
  savingsContributionDelta = 0,
}) {
  const current = calculateFinancialSnapshot({
    income, housing, food, transportation, education,
    healthcare, entertainment, subscriptions, other,
    debtPayment, savingsContribution,
  });

  const projected = calculateFinancialSnapshot({
    income: income + incomeDelta,
    housing: housing + housingDelta,
    food: food + foodDelta,
    transportation: transportation + transportationDelta,
    education: education + educationDelta,
    healthcare: healthcare + healthcareDelta,
    entertainment: entertainment + entertainmentDelta,
    subscriptions: subscriptions + subscriptionsDelta,
    other: other + otherDelta,
    debtPayment: debtPayment + debtPaymentDelta,
    savingsContribution: savingsContribution + savingsContributionDelta,
  });

  return {
    current,
    projected,
    changes: {
      income: incomeDelta,
      housing: housingDelta,
      food: foodDelta,
      transportation: transportationDelta,
      education: educationDelta,
      healthcare: healthcareDelta,
      entertainment: entertainmentDelta,
      subscriptions: subscriptionsDelta,
      other: otherDelta,
      debtPayment: debtPaymentDelta,
      savingsContribution: savingsContributionDelta,
    },
    surplusChange: projected.monthlySurplus - current.monthlySurplus,
  };
}

File: financial/calculations.py
from datetime import date


def calculate_financial_snapshot(
    income,
    housing,
    food,
    transportation,
    education,
    healthcare,
    entertainment,
    subscriptions,
    other,
    debt_payment,
    savings_contribution
):
    total_expenses = (
        housing
        + food
        + transportation
        + education
        + healthcare
        + entertainment
        + subscriptions
        + other
        + debt_payment
    )

    monthly_surplus = income - total_expenses

    savings_rate = (
        (savings_contribution / income) * 100
        if income > 0 else None
    )

    debt_to_income = (
        (debt_payment / income) * 100
        if income > 0 else None
    )

    return {
        "total_monthly_expenses": total_expenses,
        "monthly_surplus": monthly_surplus,
        "cash_flow_status": (
            "surplus" if monthly_surplus >= 0 else "deficit"
        ),
        "savings_rate": savings_rate,
        "debt_to_income_ratio": debt_to_income
    }


def calculate_goal_progress(target_amount, amount_saved):
    if target_amount <= 0:
        return {
            "remaining_amount": 0,
            "progress_percentage": 100
        }

    remaining_amount = max(target_amount - amount_saved, 0)
    progress_percentage = min(
        (amount_saved / target_amount) * 100,
        100
    )

    return {
        "remaining_amount": remaining_amount,
        "progress_percentage": progress_percentage
    }


def calculate_goal_plan(
    target_amount,
    amount_saved,
    monthly_savings,
    target_date
):
    remaining_amount = max(target_amount - amount_saved, 0)

    if remaining_amount == 0:
        return {
            "remaining_amount": 0,
            "required_monthly_saving": 0,
            "current_saving_sufficient": True
        }

    if monthly_savings <= 0:
        return {
            "remaining_amount": remaining_amount,
            "required_monthly_saving": None,
            "current_saving_sufficient": False
        }

    today = date.today()

    if isinstance(target_date, str):
        target_date = date.fromisoformat(target_date)

    months = (
        (target_date.year - today.year) * 12
        + target_date.month - today.month
    )

    if target_date.day >= today.day:
        months += 1

    if months <= 0:
        return {
            "remaining_amount": remaining_amount,
            "required_monthly_saving": None,
            "current_saving_sufficient": False
        }

    required_monthly_saving = remaining_amount / months

    return {
        "remaining_amount": remaining_amount,
        "required_monthly_saving": required_monthly_saving,
        "current_saving_sufficient": (
            monthly_savings >= required_monthly_saving
        )
    }